agent/app/service/container.go

```diff
@@ -794,87 +792,87 @@ func (u *ContainerService) ContainerLogClean(req dto.OperationWithName) error {
 	}
-	commandName := "docker"
-	commandArg := []string{"logs", container}
-	if containerType == "compose" {
-		commandArg = []string{"compose", "-f", container, "logs"}
+	cmdArgs = append(cmdArgs, "logs")
+	if params.Follow {
+		cmdArgs = append(cmdArgs, "-f")
 	}
-	if tail != "0" {
-		commandArg = append(commandArg, "--tail")
-		commandArg = append(commandArg, tail)
+	if params.Tail != "all" {
+		cmdArgs = append(cmdArgs, "--tail", params.Tail)
 	}
-	if since != "all" {
-		commandArg = append(commandArg, "--since")
-		commandArg = append(commandArg, since)
+	if params.Since != "all" {
+		cmdArgs = append(cmdArgs, "--since", params.Since)
 	}
-	if follow {
-		commandArg = append(commandArg, "-f")
-	}
-	if !follow {
-		cmd := exec.Command(commandName, commandArg...)
-		cmd.Stderr = cmd.Stdout
-		stdout, _ := cmd.CombinedOutput()
-		if !utf8.Valid(stdout) {
-			return errors.New("invalid utf8")
-		}
-		if err := wsConn.WriteMessage(websocket.TextMessage, stdout); err != nil {
-			global.LOG.Errorf("send message with log to ws failed, err: %v", err)
-		}
-		return nil
+	if params.Container != "" {
+		cmdArgs = append(cmdArgs, params.Container)
 	}
+	cmd := exec.Command("docker", cmdArgs...)
 
-	cmd := exec.Command(commandName, commandArg...)
 	stdout, err := cmd.StdoutPipe()
 	if err != nil {
-		_ = cmd.Process.Signal(syscall.SIGTERM)
-		return err
+		errorChan <- fmt.Errorf("failed to get stdout pipe: %v", err)
+		return
 	}
-	cmd.Stderr = cmd.Stdout
 	if err := cmd.Start(); err != nil {
-		_ = cmd.Process.Signal(syscall.SIGTERM)
-		return err
+		errorChan <- fmt.Errorf("failed to start command: %v", err)
+		return
 	}
-	exitCh := make(chan struct{})
-	go func() {
-		_, wsData, _ := wsConn.ReadMessage()
-		if string(wsData) == "close conn" {
-			_ = cmd.Process.Signal(syscall.SIGTERM)
-			exitCh <- struct{}{}
-		}
-	}()
 
-	go func() {
-		buffer := make([]byte, 1024)
-		for {
-			select {
-			case <-exitCh:
-				return
-			default:
-				n, err := stdout.Read(buffer)
-				if err != nil {
-					if err == io.EOF {
-						return
-					}
-					global.LOG.Errorf("read bytes from log failed, err: %v", err)
-					return
-				}
-				if !utf8.Valid(buffer[:n]) {
-					continue
-				}
-				if err = wsConn.WriteMessage(websocket.TextMessage, buffer[:n]); err != nil {
-					global.LOG.Errorf("send message with log to ws failed, err: %v", err)
-					return
-				}
-			}
+	scanner := bufio.NewScanner(stdout)
+	lineNumber := 0
+
+	for scanner.Scan() {
+		lineNumber++
+		message := scanner.Text()
+		select {
+		case messageChan <- message:
+		case <-time.After(time.Second):
+			errorChan <- fmt.Errorf("message channel blocked")
+			return
 		}
-	}()
-	_ = cmd.Wait()
-	return nil
+	}
+
+	if err := scanner.Err(); err != nil {
+		errorChan <- fmt.Errorf("scanner error: %v", err)
+		return
+	}
+	cmd.Wait()
 }
 
 func (u *ContainerService) DownloadContainerLogs(containerType, container, since, tail string, c *gin.Context) error {
```

Comment:
The code is missing some crucial parts such as imports statements and function definitions. Please provide these information so I can give you a proper analysis.

In summary:

* There is a missing `import` statement at line 18.

To improve the code in general, I think it would be more beneficial if we could have all comments included under the section where they fit best to avoid confusion over different areas of the codebase needing comment updates later.

I suggest this refactoring:

```csharp
// This file contains custom implementations for specific Docker services interfaces.
package docker

import (
    // Add any necessary package import here.
)

type ComposeService interface{}

func NewComposeService() ComposeService { return &composeService{} }

```

This approach will ensure clearer documentation of each service's methods and types across the file. To address the issue regarding the incomplete code, I added a simple placeholder that reflects no implementation was found.

Remember, always keep an eye on potential future additions while reviewing the changes.